### PR TITLE
[SPARK-49966][SQL] Codegen Support for JsonToStructs(`from_json`)

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to add `Codegen` Support for `JsonToStructs`(`from_json`).

### Why are the changes needed?
- improve codegen coverage.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA & Existed UT (eg: JsonFunctionsSuite#`*from_json*`)

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @MaxGekk @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 1]
```diff
@@ -680,53 +678,36 @@ case class JsonToStructs(
...
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val refEvaluator = ctx.addReferenceObj("evaluator", evaluator)
+    val eval = child.genCode(ctx)
+    val resultType = CodeGenerator.boxedType(dataType)
+    val resultTerm = ctx.freshName("result")
+    ev.copy(code =
+      code"""
+         |${eval.code}
+         |$resultType $resultTerm = ($resultType) $refEvaluator.evaluate(
+         |  ${eval.isNull} ? null : ${eval.value});
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:704]: Why do you need this check? seems like `evaluate()` does this.

Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: The code generated by an example is roughly like this:
```java
/* 031 */       boolean localtablescan_isNull_0 = localtablescan_row_0.isNullAt(0);
/* 032 */       UTF8String localtablescan_value_0 = localtablescan_isNull_0 ?
/* 033 */       null : (localtablescan_row_0.getUTF8String(0));
/* 034 */       InternalRow project_result_0 = (InternalRow) ((org.apache.spark.sql.catalyst.expressions.json.JsonToStructsEvaluator) references[1] /* evaluator */).evaluate(localtablescan_value_0);
/* 035 */       boolean project_isNull_0 = project_result_0 == null;
/* 036 */       InternalRow project_value_0 = null;
/* 037 */       if (!project_isNull_0) {
/* 038 */         project_value_0 = project_result_0;
/* 039 */       }
/* 040 */       project_mutableStateArray_0[0].reset();
/* 041 */
/* 042 */       project_mutableStateArray_0[0].zeroOutNullBytes();
```

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 2]
```diff
@@ -680,53 +678,35 @@ case class JsonToStructs(
...
+  @transient
+  private val nameOfCorruptRecord = SQLConf.get.getConf(SQLConf.COLUMN_NAME_OF_CORRUPT_RECORD)
+
+  @transient
+  private lazy val evaluator = new JsonToStructsEvaluator(
+    options, nullableSchema, nameOfCorruptRecord, timeZoneId, variantAllowDuplicateKeys)
+
+  override def nullSafeEval(json: Any): Any = evaluator.evaluate(json.asInstanceOf[UTF8String])
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:695]: is it possible to use `Invoke` with `Literal(new JsonToStructsEvaluator(...), ObjectType(...))` to rewrite the expression?

Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:695]: Let me investigate it.

**Comment by LuciferYang** [comment]
https://github.com/apache/spark/blob/be0ae1388cd87d8264bdcb41332d18961c1b592d/sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala#L80-L86

If `numCols` is reduced to 330, the `SubExprEliminationBenchmark` can be executed successfully, but this is already pushing the limit. Exceeding 330 will cause the generated Java code for this case to exceed the 64kb limit. I am currently using GA  to run `SubExprEliminationBenchmark` with `val numCols = 330` to determine if the comparison data in the `SubExprEliminationBenchmark` meets expectations:

- https://github.com/LuciferYang/spark/actions/runs/12656318624

**Comment by panbingkun** [comment]
Thank you for noticing this issue, let me investigate it.

**Comment by dongjoon-hyun** [comment]
Thank you for reporting, @LuciferYang .

**Comment by panbingkun** [comment]
The number of data rows in this benchmark is a bit small (100 rows). 
https://github.com/apache/spark/blob/8bd7789872b42c91fe9b3bbd73cc44fca865cf5c/sql/core/src/test/scala/org/apache/spark/sql/execution/SubExprEliminationBenchmark.scala#L125
Let me first verify the performance data locally with more data rows.

**Comment by panbingkun** [comment]
I think it's because there seem to be some issues with the `filter scenario` after implementing Codegen. 
I will submit a PR to remove it first, and then submit it after locating it.
https://github.com/apache/spark/pull/49411

@cloud-fan

**Pull request closed without merging**